text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,395 @@
+versions of `log`, instead seeing that there's an exact dependency on the git
+repository (from the `Cargo.lock`) and the repository is listed as an
+augmentation, so it'll follow that pointer.
+
+## Impact on `[replace]`
+
+The `[augment]` section in the manifest can in many ways be seen as a "replace
+2.0". It is, in fact, strictly more expressive than the current `[replace]`
+section! For example these two sections are equivalent:
+
+```toml
+[replace]
+'log:0.3.7' = { git = 'https://github.com/rust-lang-nursery/log' }
+
+# is the same as...
+
+[augment.crates-io]
+log = { git = 'https://github.com/rust-lang-nursery/log' }
+```
+
+This is not accidental! The intial development of the `[augment]` feature was
+actually focused on prepublishing dependencies and was called `[prepublish]`,
+but while discussing it a conclusion was reached that `[prepublish]` already
+allowed replacing existing versions in a registry, but issued a warning when
+doing so. It turned out that without a warning we ended up having a full-on
+`[replace]` replacement!
+
+At this time, though, it is not planned to deprecate the `[replace]` section,
+nor remove it. After the `[augment]` section is implemented, if it ends up
+working out this may change. If after a few cycles on stable the `[augment]`
+section seems to be working well we can issue an official deprecation for
+`[replace]`, printing a warning if it's still used.
+
+Documentation, however, will immediately being to recommend `[augment]` over
+`[replace]`.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+Augmentation is a feature intended for large-scale projects spanning many repos
+and crates, where you want to make something like an atomic change across the
+repos. As such, it should likely be explained in a dedicated section for
+large-scale Cargo usage, which would also include build system integration and
+other related topics.
+
+The mechanism itself is straightforward enough that a handful of examples (as in
+this RFC) is generally enough to explain it. In the docs, these examples should
+be spelled out in greater detail.
+
+Most notably, however, the [overriding dependenices][over] section of Cargo's
+documentation will be rewritten to primarily mention `[augment]`, but
+`[replace]` will be mentioned still with a recommendation to use `[augment]`
+instead if possible.
+
+[over]: http://doc.crates.io/specifying-dependencies.html#overriding-dependencies
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This feature adds yet another knob around where, exactly, Cargo is getting its
+source and version information. In particular, it's basically deprecating
+`[replace]` if it works out, and it's typically a shame to deprecate major
+stable features.
+
+Fortunately, because these features are intended to be relatively rarely used,
+checked in even more rarely, are only used for very large projects, and cannot
+be published to crates.io, the knobs are largely invisible to the vast majority
```

Comment:
Correct me if I'm wrong, but I think this is the first point in the RFC now that mentions not being able to publish crates that use `[augment]` to crates.io. I think that deserves at least a mention in the detailed design section.

Reply:
Was there a decision made here? I don't see why (at least) path-based patches _shouldn't_ be allowed. This would allow me to resolve https://github.com/SergioBenitez/Rocket/issues/513, for example.